mizar/networkpolicy/networkpolicy_util.py

```diff
@@ -30,26 +30,76 @@
+                if ep is not None:
+                    ep.remove_ingress_networkpolicy(policy_name)
+                    ep.remove_egress_networkpolicy(policy_name)
+
+        for endpoint_name in affected_endpoint_names:
+            networkpolicy_opr.store.networkpolicy_endpoints_store[policy_name].remove(endpoint_name)
+
+        for endpoint_name in endpoint_name_list:
+            networkpolicy_opr.store.add_networkpolicy_endpoint(policy_name, endpoint_name)
+            affected_endpoint_names.add(endpoint_name)
+
+        if len(networkpolicy_opr.store.networkpolicy_endpoints_store[policy_name]) == 0:
+            networkpolicy_opr.store.networkpolicy_endpoints_store.pop(policy_name)
+
+        return affected_endpoint_names
+
+    def retrieve_endpoints_for_networkpolicy(self, policy_name, pod_label_dict):
+        endpoint_name_list = set()
+
+        if pod_label_dict is None:
+            return endpoint_name_list
+
         pods = kube_list_pods_by_labels(networkpolicy_opr.core_api, pod_label_dict)
         if pods is None:
-            return
+            return endpoint_name_list
+        
+        for pod in pods.items:
+            pod_name = pod.metadata.name
+            eps = endpoint_opr.store.get_eps_in_pod(pod_name)
+            for ep in eps.values():
+                endpoint_name_list.add(ep.name)
+
+        return endpoint_name_list
 
+    def add_networkpolicy_for_endpoint(self, policy_name, policy_types):
         has_ingress = "Ingress" in policy_types
         has_egress = "Egress" in policy_types
 
         if not has_ingress and not has_egress:
             return
 
-        for pod in pods.items:
-            pod_name = pod.metadata.name
-            eps = endpoint_opr.store.get_eps_in_pod(pod_name)
-            for ep in eps.values():
-                networkpolicy_opr.store.add_networkpolicy_endpoint(name, ep.name)
-                if has_ingress and name not in ep.ingress_networkpolicies:
-                    ep.add_ingress_networkpolicy(name)
-                if has_egress and name not in ep.egress_networkpolicies:
-                    ep.add_egress_networkpolicy(name)
+        if policy_name not in networkpolicy_opr.store.networkpolicy_endpoints_store:
+            return
+        
+        for endpoint_name in networkpolicy_opr.store.networkpolicy_endpoints_store[policy_name]:
+            ep = networkpolicy_opr.store.get_ep(endpoint_name)
+            if ep is not None:
+                if has_ingress and policy_name not in ep.ingress_networkpolicies:
+                    ep.add_ingress_networkpolicy(policy_name)
+                if has_egress and policy_name not in ep.egress_networkpolicies:
+                    ep.add_egress_networkpolicy(policy_name)
+
+    def handle_networkpolicy_update_delete(self, endpoint_name_list):
+        for endpoint_name in endpoint_name_list:
+            ep = networkpolicy_opr.store.get_ep(endpoint_name)
+            if ep is not None:
```

Comment:
suggest to add logging when ep is none; my understanding is when it is none, something had happened between the ep name was got and when this func is being called, which is likey some race condition / exception.

Reply:
Fixed